src/consumer/assigners/rangeAssigner/index.js

```diff
@@ -0,0 +1,77 @@
+const { MemberMetadata, MemberAssignment } = require('../../assignerProtocol')
+
+/**
+ * RangeAssigner
+ * @type {import('types').PartitionAssigner}
+ */
+module.exports = ({ cluster }) => ({
+  name: 'RangeAssigner',
+  version: 0,
+
+  /**
+   * Assign the topics to the provided members using range strategy.
+   *
+   * The members array contains information about each member, `memberMetadata` is the result of the
+   * `protocol` operation.
+   *
+   * @param {object} group
+   * @param {import('types').GroupMember[]} group.members array of members, e.g:
+                              [{ memberId: 'test-5f93f5a3', memberMetadata: Buffer }]
+   * @param {string[]} group.topics
+   * @returns {Promise<import('types').GroupMemberAssignment[]>} object partitions per topic per member
+   */
+  async assign({ members, topics }) {
+    // logic inspired from kafka java client:
+    // https://kafka.apache.org/24/javadoc/org/apache/kafka/clients/consumer/RangeAssignor.html
+    // https://github.com/a0x8o/kafka/blob/master/clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
+
+    const sortedMembers = members.map(({ memberId }) => memberId).sort()
+    const membersCount = sortedMembers.length
+    const assignment = {}
+
+    for (const topic of topics) {
+      const partitionMetadata = cluster.findTopicPartitionMetadata(topic)
+
+      const numPartitionsForTopic = partitionMetadata.length
+      const numPartitionsPerConsumer = Math.floor(numPartitionsForTopic / membersCount)
+      const consumersWithExtraPartition = numPartitionsForTopic % membersCount
+
+      for (var i = 0; i < membersCount; i++) {
+        const start = numPartitionsPerConsumer * i + Math.min(i, consumersWithExtraPartition)
+        const length = numPartitionsPerConsumer + (i + 1 > consumersWithExtraPartition ? 0 : 1)
+
+        const assignee = sortedMembers[i]
+
+        for (let partition = start; partition < start + length; partition++) {
+          if (!assignment[assignee]) {
```

Comment:
I think we have 3 levels of for loop which doesn't seem ideal, on the other hand, this is only used once assigning members to the consumers right? So it should be fine I guess. I wanted to ask if you have reasoned about this before as well?

Reply:
Yes, same flow in java client. There isn't that much data to loop through and like you said it happens once. I see no reason to optimise.

Reply:
LGTM